addons/product/models/product.py

```diff
@@ -227,9 +228,34 @@ def init(self):
         self.env.cr.execute("CREATE UNIQUE INDEX IF NOT EXISTS product_product_combination_unique ON %s (product_tmpl_id, combination_indices) WHERE active is true"
             % self._table)
 
-    _sql_constraints = [
-        ('barcode_uniq', 'unique(barcode)', "A barcode can only be assigned to one product !"),
-    ]
+    @api.constrains("barcode")
+    def _check_barcode_unique(self):
+
+        # Collect duplicates (if there are any)
+        duplicates = defaultdict(set)
+        for product in self:
+
+            # Skip check on products without a barcode
+            if not product.barcode:
+                continue
+
+            products_with_same_barcode = self.env['product.product'].sudo().search(
```

Comment:
Please take care to not introduce performance issues.

Reply:
I understand your concern. Do you have any recommendation regarding a different course of action that would achieve the same end goal without the additional performance impact?

Reply:
IMO a good design can be: Allow inconsistencies in products (so allow barcode duplicates) and put some validation logic elsewhere, like do not allow to add a product in sale order if it's not compliant to some rules. 
An exception can be thrown at runtime or by a cron (exception like in stock for "procurement exceptions" / bad routing).

Plus here the barcode is a bit special: it's really opinionated. Only one barcode per product, unicity across all the companies, and there is edge cases when (de)archiving products. 


Back to the perf: I do not know if your code introduce a perf issue. I suspect it but I'm not an ORM specialist. My request is, before merging it, please test if there is no regressions against a db with much more than 10 products. :roll_eyes:

Reply:
I see. I will talk about this with the rest of the team. Thank you very much for your input!